web/src/add_subscribers_pill.ts

```diff
@@ -1,41 +1,79 @@
 import * as input_pill from "./input_pill";
 import * as keydown_util from "./keydown_util";
+import type {User} from "./people";
 import * as pill_typeahead from "./pill_typeahead";
 import * as stream_pill from "./stream_pill";
+import type {CombinedPillContainer} from "./typeahead_helper";
 import * as user_group_pill from "./user_group_pill";
 import * as user_pill from "./user_pill";
 
-function create_item_from_text(text, current_items) {
-    const funcs = [
-        stream_pill.create_item_from_stream_name,
-        user_group_pill.create_item_from_group_name,
-        user_pill.create_item_from_email,
-    ];
-    for (const func of funcs) {
-        const item = func(text, current_items);
-        if (item) {
+export type CombinedPillItem =
+    | input_pill.InputPillItem<user_pill.UserPill>
+    | input_pill.InputPillItem<user_group_pill.UserGroupPill>
+    | input_pill.InputPillItem<stream_pill.StreamPill>;
+
+function create_item_from_text(
+    text: string,
+    current_items: CombinedPillItem[],
+): CombinedPillItem | undefined {
+    let item: CombinedPillItem | undefined;
+    const filter_stream_pill = current_items.flatMap((item) => {
+        if (item.type === "stream") {
             return item;
         }
+        return [];
+    });
+    item = stream_pill.create_item_from_stream_name(text, filter_stream_pill);
+    if (item) {
+        return item;
     }
-    return undefined;
-}
-
-function get_text_from_item(item) {
-    const funcs = [
-        stream_pill.get_stream_name_from_item,
-        user_group_pill.get_group_name_from_item,
-        user_pill.get_email_from_item,
-    ];
-    for (const func of funcs) {
-        const text = func(item);
-        if (text) {
-            return text;
+    const filter_user_group_pill = current_items.flatMap((item) => {
+        if (item.type === "user_group") {
+            return item;
         }
+        return [];
+    });
+    item = user_group_pill.create_item_from_group_name(text, filter_user_group_pill);
+    if (item) {
+        return item;
+    }
+    const filter_user_pill = current_items.flatMap((item) => {
+        if (item.type === "user") {
+            return item;
+        }
+        return [];
+    });
+    item = user_pill.create_item_from_email(text, filter_user_pill);
+    if (item) {
+        return item;
     }
```

Comment:
It's probably worth having a #frontend thread about whether there's a better way to do this pattern, but this seems OK.

Reply:
Yes I agree. I gave a lot of thought on doing this but didn't found any better solution. I'll start the thread.